File: Fondamenti di Informatica 2/Esercitazioni con soluzioni/8.2_DFS Diretta/Soluzione/c-source/driver.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "linked_list.h"
#include "graph.h"
#include "graph_services.h"

void print_help() {
    printf("Richiesto argomento: {graph, sweep, top_sort, strong_cc, list}\n");
}

int main(int argc, char** argv) {
	if (argc != 2) {
		print_help();
		return EXIT_FAILURE;
	}
	else if (strcmp(argv[1], "list") == 0) {
		linked_list* lista = linked_list_new();
		linked_list_print(lista);
		printf("\n");

		linked_list_add(lista,0, (void*)"1");
		linked_list_print(lista);
		printf("\n");

		linked_list_add(lista,0, (void*)"2");
		linked_list_print(lista);
		printf("\n");

		linked_list_add(lista,1, (void*)"3");
		linked_list_print(lista);
		printf("\n");

		linked_list_add(lista,2, (void*)"4");
		linked_list_print(lista);
		printf("\n");

		linked_list_add(lista,4, (void*)"5");
		linked_list_print(lista);
		printf("\n");
		
		for (int i = -1; i <= linked_list_size(lista); i++) {
			printf("element in pos: %d -> %s\n", i, (char*)linked_list_get(lista, i));
		}
		printf("\n");

		linked_list_remove(lista, 0);
		linked_list_print(lista);
		printf("\n");

		linked_list_remove(lista, 0);
		linked_list_print(lista);
		printf("\n");

		linked_list_remove(lista, 1);
		linked_list_print(lista);
		printf("\n");

		linked_list_remove(lista, 1);
		linked_list_print(lista);
		printf("\n");

		linked_list_remove(lista, 0);
		linked_list_print(lista);
		printf("\n");

		linked_list_add(lista, 0,(void*) "1");
		linked_list_add(lista, 0,(void*) "2");
		linked_list_add(lista, 0,(void*) "3");
		linked_list_add(lista, 0,(void*) "5");
		linked_list_add(lista, 0,(void*) "4");
		linked_list_enqueue(lista, (void*)"4");
		linked_list_enqueue(lista, (void*)"5");
		linked_list_enqueue(lista, (void*)"1");
		linked_list_enqueue(lista, (void*)"1");
		linked_list_enqueue(lista, (void*)"4");
		linked_list_enqueue(lista, (void*)"5");

		linked_list_print(lista);
		linked_list_iterator *it = linked_list_iterator_new(lista);
		void * cur = linked_list_iterator_getvalue(it);
		for(; linked_list_iterator_hasnext(it); cur = linked_list_iterator_next(it)) {
			if (cur) {
				if (((char*)cur)[0] == '1') {
					linked_list_iterator_remove(it);
				}
				if (((char*)cur)[0] == '4') {
					linked_list_iterator_remove(it);
				}
				if (((char*)cur)[0] == '5') {
					linked_list_iterator_remove(it);
				}
			}
			//linked_list_iterator_remove(it);
			linked_list_print(lista);
		}
		linked_list_iterator_delete(it);
		
		printf("DONE\n");
		linked_list_delete(lista);
	}
	else if (strcmp(argv[1], "graph") == 0) {
		graph * graph = graph_new();
		
		graph_node * a = graph_add_node(graph, (void*)"a");
		graph_node * b = graph_add_node(graph, (void*)"b");
		graph_node * c = graph_add_node(graph, (void*)"c");
		graph_node * d = graph_add_node(graph, (void*)"d");
		graph_node * e = graph_add_node(graph, (void*)"e");
		graph_node * f = graph_add_node(graph, (void*)"f");
		graph_node * g = graph_add_node(graph, (void*)"g");
		graph_node * h = graph_add_node(graph, (void*)"h");
		graph_node * i = graph_add_node(graph, (void*)"i");

		graph_add_edge(graph, b, c);
		graph_add_edge(graph, b, f);
		graph_add_edge(graph, c, d);
		graph_add_edge(graph, d, b);
		graph_add_edge(graph, e, f);
		graph_add_edge(graph, f, c);
		graph_add_edge(graph, g, h);
		graph_add_edge(graph, a, c);
		graph_add_edge(graph, c, a);
		graph_add_edge(graph, i, a);

		printf("Grafo iniziale\n\n");
		graph_print(graph);
		printf("\n");

		printf("Rimozione nodo: %s\n", (char*)graph_get_node_value(a));
		graph_remove_node(graph, a);
		graph_print(graph);

		printf("Rimozione: %s,%s\n", (char*)graph_get_node_value(f), (char*)graph_get_node_value(c));
		graph_remove_edge(graph, f, c);
		graph_print(graph);

		printf("Rimozione: %s,%s\n", (char*)graph_get_node_value(g), (char*)graph_get_node_value(h));
		graph_remove_edge(graph, g, h);
		graph_print(graph);

		printf("Rimozione: %s,%s\n", (char*)graph_get_node_value(c), (char*)graph_get_node_value(d));
		graph_remove_edge(graph, c, d);
		graph_print(graph);

		printf("Aggiunta: %s,%s\n", (char*)graph_get_node_value(c), (char*)graph_get_node_value(f));
		graph_add_edge(graph, c, f);
		graph_print(graph);

		printf("Aggiunta: %s,%s\n", (char*)graph_get_node_value(d), (char*)graph_get_node_value(c));
		graph_add_edge(graph, d, c);
		graph_print(graph);
		printf("\n");

		printf("Rimozione nodo: %s\n", (char*)graph_get_node_value(d));
		graph_remove_node(graph, d);
		graph_print(graph);

		printf("Rimozione nodo: %s\n", (char*)graph_get_node_value(h));
		graph_remove_node(graph, h);
		graph_print(graph);

		printf("Rimozione nodo: %s\n", (char*)graph_get_node_value(i));
		graph_remove_node(graph, i);
		graph_print(graph);

		printf("Rimozione nodo: %s\n", (char*)graph_get_node_value(b));
		graph_remove_node(graph, b);
		graph_print(graph);

		printf("Rimozione nodo: %s\n", (char*)graph_get_node_value(c));
		graph_remove_node(graph, c);
		graph_print(graph);

		printf("Rimozione nodo: %s\n", (char*)graph_get_node_value(e));
		graph_remove_node(graph, e);
		graph_print(graph);

		printf("Rimozione nodo: %s\n", (char*)graph_get_node_value(f));
		graph_remove_node(graph, f);
		graph_print(graph);

		printf("Rimozione nodo: %s\n", (char*)graph_get_node_value(g));
		graph_remove_node(graph, g);
		printf("EMPTY GRAPH:\n");
		graph_print(graph);
		graph_delete(graph);
		printf("DONE\n");
	}
	else if (strcmp(argv[1], "sweep") == 0) {
		graph * graph = graph_new();

		graph_node * a = graph_add_node(graph, (void*)"a");
		graph_node * b = graph_add_node(graph, (void*)"b");
		graph_node * c = graph_add_node(graph, (void*)"c");
		graph_node * d = graph_add_node(graph, (void*)"d");
		graph_node * e = graph_add_node(graph, (void*)"e");
		graph_node * f = graph_add_node(graph, (void*)"f");
		graph_node * g = graph_add_node(graph, (void*)"g");
		graph_node * h = graph_add_node(graph, (void*)"h");

		graph_add_edge(graph, a, b);
		graph_add_edge(graph, a, f);
		graph_add_edge(graph, b, c);
		graph_add_edge(graph, b, f);
		graph_add_edge(graph, c, d);
		graph_add_edge(graph, d, b);
		graph_add_edge(graph, e, f);
		graph_add_edge(graph, f, c);
		graph_add_edge(graph, g, h);

		printf("Grafo\n\n");
		graph_print(graph);
		printf("\n");

		sweep(graph, (char*)"%s");

		graph_delete(graph);
	}
	else if (strcmp(argv[1], "top_sort") == 0) {
		graph * graph = graph_new();

		graph_node * a = graph_add_node(graph, (void*)"a");
		graph_node * b = graph_add_node(graph, (void*)"b");
		graph_node * c = graph_add_node(graph, (void*)"c");
		graph_node * d = graph_add_node(graph, (void*)"d");
		graph_node * e = graph_add_node(graph, (void*)"e");
		graph_node * f = graph_add_node(graph, (void*)"f");
		graph_node * g = graph_add_node(graph, (void*)"g");
		graph_node * h = graph_add_node(graph, (void*)"h");

		graph_add_edge(graph, a, b);
		graph_add_edge(graph, a, f);
		graph_add_edge(graph, b, c);
		graph_add_edge(graph, b, f);
		graph_add_edge(graph, c, d);
		graph_add_edge(graph, d, b);
		graph_add_edge(graph, e, f);
		graph_add_edge(graph, f, c);
		graph_add_edge(graph, g, h);

		printf("Grafo\n\n");
		graph_print(graph);
		printf("\n");

		printf("Primo tentativo: (dovrebbe fallire)\n");
		topological_sort(graph);

		printf("Rimozione arco (d,b)\n");
		graph_remove_edge(graph, d, b);

		printf("secondo tentativo: (dovrebbe riuscire)\n");
		topological_sort(graph);

		graph_delete(graph);
	}
	else if (strcmp(argv[1], "strong_cc") == 0) {
		graph * graph = graph_new();

		graph_node * a = graph_add_node(graph, (void*)"a");
		graph_node * b = graph_add_node(graph, (void*)"b");
		graph_node * c = graph_add_node(graph, (void*)"c");
		graph_node * d = graph_add_node(graph, (void*)"d");
		graph_node * e = graph_add_node(graph, (void*)"e");

		graph_add_edge(graph, a, c);
		graph_add_edge(graph, a, d);
		graph_add_edge(graph, b, a);
		graph_add_edge(graph, d, b);
		graph_add_edge(graph, e, c);
		
		printf("Grafo 1\n\n");
		graph_print(graph);
		printf("\n");

		strong_connected_components(graph);

		graph_delete(graph);
		
		graph = graph_new();

		a = graph_add_node(graph, (void*)"a");
		b = graph_add_node(graph, (void*)"b");
		c = graph_add_node(graph, (void*)"c");
		d = graph_add_node(graph, (void*)"d");
		e = graph_add_node(graph, (void*)"e");
		graph_node * f = graph_add_node(graph, (void*)"f");
		graph_node * g = graph_add_node(graph, (void*)"g");
		graph_node * h = graph_add_node(graph, (void*)"h");

		graph_add_edge(graph, a, b);
		graph_add_edge(graph, a, f);
		graph_add_edge(graph, b, c);
		graph_add_edge(graph, b, f);
		graph_add_edge(graph, c, d);
		graph_add_edge(graph, d, b);
		graph_add_edge(graph, e, f);
		graph_add_edge(graph, f, c);
		graph_add_edge(graph, g, h);
		
		printf("Grafo 2\n\n");
		graph_print(graph);
		printf("\n");

		strong_connected_components(graph);

		graph_delete(graph);

		graph = graph_new();

		a = graph_add_node(graph, (void*)"0");
		b = graph_add_node(graph, (void*)"1");
		c = graph_add_node(graph, (void*)"2");
		d = graph_add_node(graph, (void*)"3");
		e = graph_add_node(graph, (void*)"4");
		f = graph_add_node(graph, (void*)"5");
		g = graph_add_node(graph, (void*)"6");
		h = graph_add_node(graph, (void*)"7");
		graph_node * i = graph_add_node(graph, (void*)"8");

		// 0-1-2-3
		graph_add_edge(graph, a, b);
		graph_add_edge(graph, b, c);
		graph_add_edge(graph, c, d);
		graph_add_edge(graph, d, a);

		graph_add_edge(graph, c, e);

		// 4-5-6
		graph_add_edge(graph, e, f);
		graph_add_edge(graph, f, g);
		graph_add_edge(graph, g, e);

		graph_add_edge(graph, h, g);
		graph_add_edge(graph, h, i);


		
		printf("Grafo 3\n\n");
		graph_print(graph);
		printf("\n");

		strong_connected_components(graph);

		graph_delete(graph);
		
	}
	else{
        print_help();
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

File: Fondamenti di Informatica 2/Esercitazioni con soluzioni/8.2_DFS Diretta/distibuzione/java-aux/Driver.java
public class Driver {

    public static void print() {
        System.out.println("Richiesto argomento: {graph, sweep, top_sort, strong_cc}");
    }

    public static void main(String[] argv) {

        if (argv.length < 1) {
            print();
            return;
        }
        
        else if (argv[0].equals("graph")) {
            Graph<String> graph = new Graph<String>();

            Graph.GraphNode<String> a = graph.addNode("a");
            Graph.GraphNode<String> b = graph.addNode("b");
            Graph.GraphNode<String> c = graph.addNode("c");
            Graph.GraphNode<String> d = graph.addNode("d");
            Graph.GraphNode<String> e = graph.addNode("e");
            Graph.GraphNode<String> f = graph.addNode("f");
            Graph.GraphNode<String> g = graph.addNode("g");
            Graph.GraphNode<String> h = graph.addNode("h");

            graph.addEdge(a, b);
            graph.addEdge(a, f);
            graph.addEdge(b, c);
            graph.addEdge(b, f);
            graph.addEdge(c, d);
            graph.addEdge(d, b);
            graph.addEdge(e, f);
            graph.addEdge(f, c);
            graph.addEdge(g, h);



            System.out.println("Grafo iniziale");
            //System.out.println(graph.getNodes());
            System.out.println(graph);

            System.out.println("Rimozione " + f + "," + c);
            graph.removeEdge(f,c);
            System.out.println(graph);

            System.out.println("Rimozione " + c + "," + b);
            graph.removeEdge(c,b);
            System.out.println(graph);

            System.out.println("Rimozione " + c + "," + d);
            graph.removeEdge(c,d);
            System.out.println(graph);

            System.out.println("Aggiunta " + c + "," + f);
            graph.addEdge(c,f);
            System.out.println(graph);

            System.out.println("Aggiunta " + d + "," + c);
            graph.addEdge(d,c);
            System.out.println(graph);

            System.out.println("Rimozione nodo " + d);
            graph.removeNode(d);
            System.out.println(graph);

            System.out.println("Rimozione nodo " + f);
            graph.removeNode(f);
            System.out.println(graph);

            System.out.println("Rimozione nodo " + c);
            graph.removeNode(c);
            System.out.println(graph);
        }
        else if (argv[0].equals("sweep")) {
        	Graph<String> graph = new Graph<String>();

            Graph.GraphNode<String> a = graph.addNode("a");
            Graph.GraphNode<String> b = graph.addNode("b");
            Graph.GraphNode<String> c = graph.addNode("c");
            Graph.GraphNode<String> d = graph.addNode("d");
            Graph.GraphNode<String> e = graph.addNode("e");
            Graph.GraphNode<String> f = graph.addNode("f");
            Graph.GraphNode<String> g = graph.addNode("g");
            Graph.GraphNode<String> h = graph.addNode("h");

            graph.addEdge(a, b);
            graph.addEdge(a, f);
            graph.addEdge(b, c);
            graph.addEdge(b, f);
            graph.addEdge(c, d);
            graph.addEdge(d, b);
            graph.addEdge(e, f);
            graph.addEdge(f, c);
            graph.addEdge(g, h);
            
            System.out.println("Grafo:");
            System.out.println(graph);
            GraphServices.sweep(graph);
        }
        else if (argv[0].equals("top_sort")) {
        	Graph<String> graph = new Graph<String>();

            Graph.GraphNode<String> a = graph.addNode("a");
            Graph.GraphNode<String> b = graph.addNode("b");
            Graph.GraphNode<String> c = graph.addNode("c");
            Graph.GraphNode<String> d = graph.addNode("d");
            Graph.GraphNode<String> e = graph.addNode("e");
            Graph.GraphNode<String> f = graph.addNode("f");
            Graph.GraphNode<String> g = graph.addNode("g");
            Graph.GraphNode<String> h = graph.addNode("h");

            graph.addEdge(a, b);
            graph.addEdge(a, f);
            graph.addEdge(b, c);
            graph.addEdge(b, f);
            graph.addEdge(c, d);
            graph.addEdge(d, b);
            graph.addEdge(e, f);
            graph.addEdge(f, c);
            graph.addEdge(g, h);
            
            System.out.println("Grafo:");
            System.out.println(graph);
            
            System.out.println("Primo tentativo: (dovrebbe fallire)");
            GraphServices.topologicalSort(graph);
            
            System.out.println("Rimozione arco (d,b)");
            graph.removeEdge(d, b);
            
            System.out.println("Nuovo Grafo:");
            System.out.println(graph);
            
            System.out.println("Secondo tentativo: (dovrebbe riuscire)");
            GraphServices.topologicalSort(graph);
        }
        else if (argv[0].equals("strong_cc")) {
        	Graph<String> graph = new Graph<String>();

            Graph.GraphNode<String> a = graph.addNode("a");
            Graph.GraphNode<String> b = graph.addNode("b");
            Graph.GraphNode<String> c = graph.addNode("c");
            Graph.GraphNode<String> d = graph.addNode("d");
            Graph.GraphNode<String> e = graph.addNode("e");
            Graph.GraphNode<String> f = graph.addNode("f");
            Graph.GraphNode<String> g = graph.addNode("g");
            Graph.GraphNode<String> h = graph.addNode("h");

            graph.addEdge(a, b);
            graph.addEdge(a, f);
            graph.addEdge(b, c);
            graph.addEdge(b, f);
            graph.addEdge(c, d);
            graph.addEdge(d, b);
            graph.addEdge(e, f);
            graph.addEdge(f, c);
            graph.addEdge(g, h);
            
            System.out.println("Grafo 1:");
            System.out.println(graph);
            
            GraphServices.strongConnectedComponents(graph);
            System.out.println("\n");

            Graph<String> graph2 = new Graph<String>();

            Graph.GraphNode<String> a2 = graph2.addNode("a");
            Graph.GraphNode<String> b2 = graph2.addNode("b");
            Graph.GraphNode<String> c2 = graph2.addNode("c");
            Graph.GraphNode<String> d2 = graph2.addNode("d");
            Graph.GraphNode<String> e2 = graph2.addNode("e");

            graph2.addEdge(a2, b2);
            graph2.addEdge(b2, c2);
            graph2.addEdge(c2, a2);
            graph2.addEdge(a2, d2);
            graph2.addEdge(d2, e2);
            
            System.out.println("Grafo 2:");
            System.out.println(graph2);
            
            GraphServices.strongConnectedComponents(graph2);
            System.out.println("\n");

            Graph<String> graph3 = new Graph<String>();

            Graph.GraphNode<String> a3 = graph3.addNode("0");
            Graph.GraphNode<String> b3 = graph3.addNode("1");
            Graph.GraphNode<String> c3 = graph3.addNode("2");
            Graph.GraphNode<String> d3 = graph3.addNode("3");
            Graph.GraphNode<String> e3 = graph3.addNode("4");
            Graph.GraphNode<String> f3 = graph3.addNode("5");
            Graph.GraphNode<String> g3 = graph3.addNode("6");
            Graph.GraphNode<String> h3 = graph3.addNode("7");
            Graph.GraphNode<String> i3 = graph3.addNode("8");

            // 0-1-2-3
            graph3.addEdge(a3, b3);
            graph3.addEdge(b3, c3);
            graph3.addEdge(c3, d3);
            graph3.addEdge(d3, a3);

            graph3.addEdge(c3, e3);

            // 4-5-6
            graph3.addEdge(e3, f3);
            graph3.addEdge(f3, g3);
            graph3.addEdge(g3, e3);

            graph3.addEdge(h3, g3);
            graph3.addEdge(h3, i3);


            
            System.out.println("Grafo 3:");
            System.out.println(graph3);
            
            GraphServices.strongConnectedComponents(graph3);
            System.out.println("\n");
        }
    } //main
} // class
